homeworks/solutions/torantie/3-smooth/README.md

```diff
@@ -0,0 +1,97 @@
+# Homework 3: Laplacian Smoothing
+
+In this homework, your task is to implement the Laplacian smoothing algorithm
+for the **uniform** and the **cotan** Laplace-Beltrami operator.
+
+## Skeleton
+
+The skeleton for this homework is structured as follows:
+
+```
+3-smooth
+├── package.json
+├── package-lock.json
+├── README.md             <--- You work on this file
+├── src
+│   ├── assets
+│   │   └── bunny_tri.obj
+│   ├── halfedge.js       <--- You work on this file
+│   ├── main.js
+│   ├── renderer.js
+│   └── vec.js
+└── webpack.config.js
+```
+
+You coding tasks are located in the `halfedge.js`.
+Please look for `TODO:` in these two files and complete them to archive
+the reference results listed below.
+
+> It might be useful to read the code in `main.js`.
+
+You should also document your implementation process. Specifically, you
+should answer the following questions:
+
+1. **Implementation complexity**: Which code snippet (report in line numbers) in the `halfedge.js` is the most time consuming for you to implement? Explain your coding experience and encountered challenges briefly.
+
+```
+TODO: your answer goes here
+```
+
+line 562-631 the smooth function: Translating the linear equation into code was very time consuming because i didn't understand or find what some things of the equations ment in the slides. Also debugging the Matrices was extremely hard because you couldn't print sparse matrices.
+
+2. **Debugging complexity**: Describe an impressive bug that you wrote while implementing this project, and briefly explain how you fixed it.
+
+```
+TODO: your answer goes here
+```
+
+I worked with the already changed array of vertices and not with the original vertice positions, resulting in incorrect calculations. Also after one smoothing operation the vertice positions were not reset with the original positions, therefore i couldn't do multiple smoothing operations. I fixed it by filling the originalVertices array with the original vertices and their original positions.
```

Comment:
This is one of the anticipated bugs :)

In my reference solution, I did a few hacks when constructing the halfedge representation. So far, our solutions both work just fine. However, it might not be ideal or elegant.

As a further question after the implementation, you could think about: how to design a data structure that enables structure backups/versioning with minimum memory duplications?